redskyctl/internal/commander/reader.go

```diff
@@ -0,0 +1,177 @@
+func NewResourceReader() *ResourceReader {
+	rr := &ResourceReader{
+		Scheme:           runtime.NewScheme(),
+		PreferredVersion: OnlyVersion,
+	}
+
+	// Always add our types
+	_ = redskyv1beta1.AddToScheme(rr.Scheme)
+	_ = redskyv1alpha1.AddToScheme(rr.Scheme)
+
+	// Allow single experiments to target an experiment list
+	_ = addExperimentListConversions(rr.Scheme)
+
+	return rr
+}
+
+// ReadInto decodes the supplied byte stream into the target runtime object. The default values
+// and type information of the target object will be populated.
+func (r *ResourceReader) ReadInto(reader io.ReadCloser, target runtime.Object) error {
+	// Determine the GVK for the type we are supposed to be populating
+	gvk, err := r.objectKind(target)
+	if err != nil {
+		return err
+	}
+
+	mt := mediaType(reader)
+	decoder, err := r.decoder(mt)
+	if err != nil {
+		return err
+	}
+
+	// TODO For lists we should consider yaml.NewDocumentDecoder(reader) so we can read a stream
+
+	// Read in the raw bytes
+	defer func() { _ = reader.Close() }()
+	data, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return err
+	}
+
+	// Decode the raw data
+	obj, _, err := decoder.Decode(data, &gvk, target)
+	if err != nil {
+		return err
+	}
+
+	// If decode returned an object of a different type we should try to force the conversion
+	if obj != target {
+		if err := r.Scheme.Convert(obj, target, r.PreferredVersion); err != nil {
+			return err
+		}
+	}
+
+	// Fill in the default values for the target
+	target.GetObjectKind().SetGroupVersionKind(gvk)
+	r.Scheme.Default(target)
+
+	return nil
+}
+
+func (r *ResourceReader) objectKind(obj runtime.Object) (schema.GroupVersionKind, error) {
+	gvks, _, err := r.Scheme.ObjectKinds(obj)
+	if err != nil {
+		return schema.GroupVersionKind{}, err
+	}
+
+	gvk, ok := r.PreferredVersion.KindForGroupVersionKinds(gvks)
+	if !ok {
+		// Your code must supply a GroupVersioner to disambiguate this case
+		panic("read destination type is ambiguous in scheme")
+	}
+
+	return gvk, nil
+}
+
+func (r *ResourceReader) decoder(mediaType string) (runtime.Decoder, error) {
+	// Create the conversion serializer with our negotiation logic in it
+	cs := controller.NewConversionSerializer(r.Scheme)
+	info, ok := runtime.SerializerInfoForMediaType(cs.SupportedMediaTypes(), mediaType)
+	if !ok {
+		return nil, fmt.Errorf("could not find serializer for %s", mediaType)
+	}
+
+	return cs.DecoderToVersion(info.Serializer, r.PreferredVersion), nil
+}
+
+// mediaType returns the presumed media type for the supplied read closer.
+func mediaType(r io.ReadCloser) string {
+	// For now just assume YAML unless we got a JSON file
+	mt := runtime.ContentTypeYAML
+	if f, ok := r.(*os.File); ok {
+		switch filepath.Ext(f.Name()) {
+		case "json":
+			mt = runtime.ContentTypeJSON
+		}
+	}
+
+	return mt
+}
+
+// OnlyVersion is a group version that only resolves if there is a single possible kind
+var OnlyVersion runtime.GroupVersioner = onlyVersion{}
```

Comment:
Can you expand on the usefulness of this?

Reply:
A group versioner is used to disambiguate the case when you register the same Go type under multiple GVKs. There are two implementations provider: disabled and internal. Disabled just returns false (which, now that I think about might be good enough if it only gets called if the `len(gvks) > 1`). The internal implementation looks for a special group name and then falls back to the first registered GVK (for some unknown definition of "first").

This group versioner basically says "if you registered the same Go object into the scheme twice, you need to bring your own `GroupVersioner` or else we panic.

Reply:
It looks like the existing "disabled" implementation won't quite work.